### DISCUSSION TITLE
Potential bug moving from java-client 8.5.0 to 8.5.1?

### DISCUSSION THREAD

**Discussion by Auto81** [appium/appium#18860] [Q&A]
Working on getting a minimal reproducible solution, but in the mean time...

I have updated my java-client from version 8.5.0 to 8.5.1, and test run times have exploded. I can swap these versions back and forth with the same results happening.

8.5.1 seems to get progressively slower. As you can see each test in the same suite taking longer and longer using 8.5.1. It's the same test setup to run back to back 10 times so it should be roughly the same times for each result.

GitHub seems to be having issues uploading images currently.

Can't use Java 17 with 8.5.0, so both results below are using Java 11. 

| test | 8.5.0  | 8.5.1 |
| ------------| ------------- | ------------- |
| 1 | 20 seconds | 33 seconds  |
| 2 | 17 seconds | 34 seconds  |
| 3 | 17 seconds | 38 seconds  |
| 4 | 17 seconds | 51 seconds  |
| 5| 18 seconds | 75 seconds  |
| 6 | 17 seconds | 113 seconds  |
| 7| 18 seconds | 162 seconds  |
| 8 | 18 seconds | 221 seconds  |
| 9| 17 seconds | 287 seconds  |
| 10| 17 seconds | 366 seconds  |

**Comment by jlipps** [comment]
Do you notice any differences in the Appium log (run with `--log-timestamp`) between sessions executed with these two java client versions?

**Reply by Auto81** [reply]
Proving tricky to get a small/neat reproducible project.

Doesn't look to be causing problems with our native app tests, which is great.

However, we are using the Appium Field Decorator in some of our pages/PageFactory as part of our selenium web tests. So might be something conflicting when combining them that way?

 Will get a reproducible pulled together asap

**Reply by Auto81** [reply]
@jlipps 

Repro Project here - [Demo](https://github.com/Auto81/AppiumFieldDecorator)

Specific test here - [Test](https://github.com/Auto81/AppiumFieldDecorator/blob/main/src/test/java/AppiumFieldDecoratorTests.java)

Pages inited, time taken in seconds, and a crude average time per POM
Not Using Appium Field Decorator (Commented out in ChallengingDomPage)
| Iterations | 8.5.0               | 8.5.1               |
|----|----|----|
| 1000       | 10 (10.0ms per POM) | 10 (10.0ms per POM) |
| 2000       | 17 (08.5ms per POM) | 17 (08.5ms per POM) |
| 3000       | 22 (07.3ms per POM) | 22 (07.3ms per POM) |

Pages inited, time taken in seconds, and a crude average time per POM
Using Appium Field Decorator
| Iterations | 8.5.0               | 8.5.1                |
|----|----|----|
| 1000       | 11 (11.0ms per POM) | 19  (19.0ms per POM) |
| 2000       | 17 (08.5ms per POM) | 46  (23.0ms per POM) |
| 3000       | 22 (07.3ms per POM) | 106 (35.3ms per POM) |
| 4000       | 30 (07.5ms per POM) | 240 (60.0ms per POM) |
| 5000       | 38 (07.6ms per POM) | 457 (91.4ms per POM) |

**Reply by jlipps** [reply]
Probably an actual java client contributor like @mykola-mokhnach or @SrinivasanTarget should investigate this

**Reply by mykola-mokhnach** [reply]
The main change there was https://github.com/appium/java-client/pull/1923

I assume cglib performs faster that bytebuddy, although I don't see any way to keep it as it has known (possibly unresolvable) compatibility issues with newer java versions.

If someone knows how to improve the actual performance there then feel free to propose a PR

**Reply by Auto81** [reply]
@mykola-mokhnach It looks to be the [ProxyListenersContainer](https://github.com/appium/java-client/pull/1923/files#diff-cdea7e4dd6e5f62229d30f69d73f9595eb6b1e058c0381d09d9b60f803502598)

I see the new (and useful) comment about changing from WeakHashMap to List. The List (8.5.1) constantly grows with each new POM/Listener, thus taking longer to loop through each time. The 'wasInstancePresent' flag isn't triggering so it's always adding a new Pair to the list. 

Would you expect this to be the case, or should the 'wasInstancePresent'  be being triggered on the second new POM etc

**Reply by mykola-mokhnach** [reply]
> The List (8.5.1) constantly grows with each new POM/Listener, thus taking longer to loop through each time

Have you tried to debug your code and check why the list has so many objects? The current implementation has WeakRef, which means it only removes instances from the static list when they are ready to be garbage-collected. Are you sure there are no extra entities in your code that may still keep references to old proxies thus preventing them from being garbage-collected?

**Reply by Auto81** [reply]
We use Cucumber and PicoContainer. By the looks of it there's a new [version 7.13.0](https://github.com/cucumber/cucumber-jvm/releases/tag/v7.13.0) that mentions a fix for 'Corrected the pico-container lifecycle: the start, stop and dispose methods are now called for every test scenarios.'

I'm running our test suite currently with that update, hopefully fixing the issue

**Reply by Auto81** [reply]
7.13.0 hasn't fixed the issue. I'll try and debug through to see why 8.5.1 is growing, but 8.5.0 is ok.

**Reply by Auto81** [reply]
Have managed to make as basic as possible project. [Project here](https://github.com/Auto81/appiumListeners851)

Basically 5 separate JUnit tests. The appium Listener size should be being Garbage collected between tests.
    
When widgets are provided via 'new' on a Page Object Model, And PageFactory.initElements() in the widget, the listeners get cleaned up as expected.

However when widgets are provided via FindBy Annotations on a Page Object Model, Some listeners are hanging over between test causing slow down eventually due to how listeners are iterated though.

There are more comments in the example project attached.

If there's anything else I can do to help assess this please let me know

**Reply by mykola-mokhnach** [reply]
This most likely means there is a memory leak somewhere, e. g. references are not being released when they are supposed to be released. 
Taking into account the downstream source is pretty complicated and I don't have much time to debug it, it would be cool if someone could help checking that.

**Reply by mykola-mokhnach** [reply]
@Auto81 I have quickly prepared https://github.com/appium/java-client/pull/1981

Could you please test this branch with your data and see if it helps?

**Reply by Auto81** [reply]
Hey, thanks for a quick build!

TopPage -> new Widget(element) -> pageFactory.initElement(appiumFieldDecorator) :: Still works as before by the looks of it

TopPage -> @FindBy Widget -> :: fails when trying to call/decorate an element on the widget, with the below exception

```
java.lang.NullPointerException: Cannot invoke "java.lang.ref.WeakReference.get()" because "this.cachedElementReference" is null
	at io.appium.java_client.pagefactory.WidgetInterceptor.call(WidgetInterceptor.java:95)
	at io.appium.java_client.proxy.Interceptor.intercept(Interceptor.java:75)
	at helloListeners.pages.widgets.ContentWidget$ByteBuddy$9NT12qcp.getParagraphText(Unknown Source)
	at helloListeners.JunitRunnerTests.code(JunitRunnerTests.java:93)

```

**Reply by mykola-mokhnach** [reply]
fixed the above exception

**Reply by Auto81** [reply]
TopPage -> new Widget(element) -> pageFactory.initElement(appiumFieldDecorator) :: Still works as expected by the looks of it


TopPage -> @findby Widget -> @findby element :: still causing hangover between test cases

**Reply by mykola-mokhnach** [reply]
:(

I've put cached widgets into WeakHashMap. Maybe this would help

**Reply by Auto81** [reply]
TopPage -> new Widget(element) -> pageFactory.initElement(appiumFieldDecorator) :: Still works as expected by the looks of it

TopPage -> @findby Widget -> @findby element :: still causing hangover between test cases

**Reply by mykola-mokhnach** [reply]
Unfortunately I don't have more time to play with the PR now.
Thanks for spending time on checking that.

**Reply by Auto81** [reply]
No problem, happy to help, and thanks for the effort

Just for notes/observations for any other attempts by anyone:

The only listeners that seem to be hanging over are ones from the widgets, when the widget itself is from a Findby 

Given a Page has a @FindBy Widget 
And the Widget has 2 @FindBy elements (eg paragraph1, paragraph2)
When a test requests either of them (eg widget.getParagraph1Text())
Then there will be 2 listeners that hangover

Reports as below for the different styles when ran through [example project here.](https://github.com/Auto81/appiumListeners851/tree/857c034fb954fb02461b7bd7018d972020225e14)

[FindBy Annotations showing hangover - Junit report](https://raw.githack.com/Auto81/appiumListeners851/857c034fb954fb02461b7bd7018d972020225e14/AnnotatedWidgetJunitRunnerTests.html)

[New widget() not hanging-over - Junit report](https://raw.githack.com/Auto81/appiumListeners851/857c034fb954fb02461b7bd7018d972020225e14/NewWidgetJunitRunnerTests.html )

**Reply by mykola-mokhnach** [reply]
It looks like there was some issue, so my recent commit has not been pushed properly to the remote repo. Please pull the branch and check it once more.

**Reply by Auto81** [reply]
Latest update on branch I can see is 14 hours ago

**Reply by mykola-mokhnach** [reply]
how about now? Is the recent update visible?

**Reply by Auto81** [reply]
Yes looks updated now, will run it this evening :)

**Reply by Auto81** [reply]
That one is very close!

TopPage -> new Widget(element) -> pageFactory.initElement(appiumFieldDecorator) :: Still works as expected by the looks of it

TopPage -> @findby Widget -> @findby element :: 50ish% of the time, it will clean up the listeners which is awesome, HOWEVER... the other half of the time it will throw a null pointer exception as follows. This is when the elements within the widget are being accessed

```
java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because the return value of "java.util.Map.get(Object)" is null

	at io.appium.java_client.pagefactory.WidgetInterceptor.getObject(WidgetInterceptor.java:92)
	at io.appium.java_client.pagefactory.interceptors.InterceptorOfASingleElement.call(InterceptorOfASingleElement.java:65)
	at io.appium.java_client.pagefactory.WidgetInterceptor.call(WidgetInterceptor.java:103)
	at io.appium.java_client.proxy.Interceptor.intercept(Interceptor.java:75)
	at helloListeners.pages.widgets.ContentWidget$ByteBuddy$6C1oUIif.getParagraph1Text(Unknown Source)
```

**Reply by mykola-mokhnach** [reply]
this sounds better. 
I've changed widget referenced to hard ones. Lets see how it behaves now

**Reply by Auto81** [reply]
This also look promising on the example project!

I'm going to plug it into my main project here and run it through. Should be able to let you know by end of day. (GMT here)

**Reply by Auto81** [reply]
Only got this kicked off late in the day, but early signs look good so far

**Reply by Auto81** [reply]
Run completed, which 8.5.1 couldn't do, so I'm gonna say it's good to go :D

Similar times to 8.5.0, and should let us move up to jdk17 now. Big win before the weekend!

Thanks so much for the effort on this, really appreciated

**Reply by mykola-mokhnach** [reply]
Thank you for checking it. This one was not simple at all.
Maybe you could also verify how lists of widgets are behaving if you can?

**Reply by Auto81** [reply]
No problem at all, glad to help

The suite I ran has the below combos, so I think it covers everything at some point. 1000~ BDD scenarios, with roughly/atleast 5 Page Objects for each. So hopefully that'd be enough to flush anything out. I'll update the smaller test project above just to be sure though

```
- Pages with WebElements, Widgets, List<WebElements>, List<Widgets>
- Widgets with WebElements, Widgets, List<WebElements>, List<Widgets>
```

**Reply by Auto81** [reply]
Smaller project with those various combinations look happy too

**Reply by Auto81** [reply]
@mykola-mokhnach I see the PR has been completed and in master now, is there a release schedule/timeframe for java-client 8.5.2/8.6.0?

**Comment by olezvinskyi** [comment]
Hi I am facing a memory leak with version 8.5.1.

 Caused by: java.lang.OutOfMemoryError: Java heap space

Java 15
Appium 2

[7_Run Gradle Script.txt.zip](https://github.com/appium/appium/files/12342013/7_Run.Gradle.Script.txt.zip)

**Comment by rafamurer** [comment]
I am facing the same issue!
Currently I have it working fine with jdk-17
`<appiumJavaClient.version>8.5.0</appiumJavaClient.version>`
forcing `<selenium.version>4.13.0</selenium.version>`

if I try to use any appium java client with higher version **8.5.1, 8.6.0 or 9.0.0** the last half of my 200 tests always takes much more time.

using Page object Model with 
`PageFactory.initElements(new AppiumFieldDecorator(this.driver), this);`
and all elements with Appium Field Decorator like that
```
  @AndroidFindBy(xpath = "xpath for android")
  @iOSXCUITFindBy(accessibility = "accessb id for iOS")
  private WebElement newInOnlineShopTitle;
```

could please advise how to update to latest appium java client without this memory leak?
should I consider java 21?

**Reply by Auto81** [reply]
Have you tried the master branch yet? Instructions [here](https://github.com/appium/java-client?tab=readme-ov-file#maven-1)

**Reply by rafamurer** [reply]
I did add in my pom file
```
    <repositories>
        <repository>
            <id>jitpack.io</id>
            <url>https://jitpack.io</url>
        </repository>
    </repositories>
```
```
        <dependency>
            <groupId>com.github.appium</groupId>
            <artifactId>java-client</artifactId>
            <version>43bbd488ffafe43eb354a7df1a031b72aeea2043</version>
        </dependency>
```

I using a cloud device provider that I can easely see this "delays"
with java client 8.5.0
<img width="383" alt="image" src="https://github.com/appium/appium/assets/36237795/58cbaa47-2589-42cb-bc04-ec3b6547c52c">
EXPECTED


and with java client 8.5.1, 8.6.0, 9.0.0 or latest master we have this

<img width="429" alt="image" src="https://github.com/appium/appium/assets/36237795/5dbd5cd6-5fc1-40a3-9d4f-4d77f3e40c42">

ISSUE

**Comment by amedvedjev** [comment]
Same for me. Same tried Java 11 and Java 17 have now. Stuck on 8.5.0 version.

https://discuss.appium.io/t/appium-terminating-unexpectedly-not-receiving-commands/40782/12

**Comment by amedvedjev** [comment]
java-client: 9.1.0 still reproducing
![image](https://github.com/appium/appium/assets/2756856/c21d6ae7-f205-4804-a64b-c7f8e3faf5b1)

**Comment by rafamurer** [comment]
Hi, is there a way we can follow up this issue? I feel like I seeing other issues that might resolve if I update the java client which I cannot due to this issue.
same discussion
https://discuss.appium.io/t/appium-terminating-unexpectedly-not-receiving-commands/40782/13

**Reply by rafamurer** [reply]
This was fixed! I tested with ` <version>9.2.3</version>`
